### DISCUSSION TITLE
Migration session's deadlock priority

### DISCUSSION THREAD

**Discussion by Avithohol** [fluentmigrator/fluentmigrator#1567] [Ideas] [answered]
What would be the best way to set DEADLOCK_PRIORITY for the migrations' sessions?

It would be useful in one database, but multi schema environments, where Fluentmigrator is run automatically, creates schemas, then migrates them separately (dedicated VersionInfo table in each schema works just fine).

In such environment database level schema stability locks may be issued by the migration DDLs, so the best is to go with the shortest transaction times for the migrations (TransactionBehavior.None).
But still, I can observe migrations get deadlocks with users' queries, or other schema migrations.

It would be nice to state the all migrations have low deadlock priority, so at least they "loose" against users' queries,
and in my Fluentmigrator migrator app, I can employ retry logic if it gets deadlocked from other migrations, so eventually all migration instances get through.

**Comment by jzabroski** [comment, marked as answer]
@Avithohol Respectfully, I think you are going down a rabbit hole. I would never allow multiple database administrators to make changes to the database simultaneously.  I could explain how to do what you want to do, but I would **never** let anyone on my team **ever** do this.

I'm very sorry.  I try to keep an open mind, and you are spot on that doing DDL operations will lock the database's system dictionaries.

I think you need to first solve the problem where you don't have two schema migrations running concurrently, then we can talk about migrations interfering with user queries.  Those are two different questions.  One is simply a terrible idea.  The other likely implies you may need to think about the problem differently.  Either the user queries need rethinking or the migrations might be trying to do too much in one go.

At very large scales, I have heard very very high-end SQL Server consultants use some fancy NAS tricks to update one copy of the database and take advantage of the fact that are read-only filegroup wont be updated, and then swap the mdf's on the real database once you want to go full live - such migrations I have heard take a week-long process at some Fortune 500 companies.  If you're at this scale, hire one of those consultants.

**Reply by Avithohol** [reply]
Thank you for the answer. I know, it sounds like a big stretch, but automatic schema creators and migrations are all the rage around here.

Regarding issue #1, the parallel migrations of independent schemas, all in all, the TransactionBehavior.None works great as even complex migrations take short locks on dictionary tables. Until no two migrations are happening on the same schema it is good.

Regarding #2, the clients vs migrations, right, there is no easy way around then to find proper maintenance windows.
